src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Events/OpenIdConnectEvents.cs

```diff
@@ -39,17 +29,22 @@ public class OpenIdConnectEvents : RemoteAuthenticationEvents, IOpenIdConnectEve
         /// <summary>
         /// Invoked before redirecting to the identity provider to authenticate.
         /// </summary>
-        public Func<RedirectContext, Task> OnRedirectToAuthenticationEndpoint { get; set; } = context => Task.FromResult(0);
+        public Func<RedirectContext, Task> OnRedirectToIdentityProvider { get; set; } = context => Task.FromResult(0);
 
         /// <summary>
         /// Invoked before redirecting to the identity provider to sign out.
         /// </summary>
-        public Func<RedirectContext, Task> OnRedirectToEndSessionEndpoint { get; set; } = context => Task.FromResult(0);
+        public Func<RedirectContext, Task> OnRedirectToIdentityProviderForSignOut { get; set; } = context => Task.FromResult(0);
 
         /// <summary>
         /// Invoked after "authorization code" is redeemed for tokens at the token endpoint.
         /// </summary>
-        public Func<TokenResponseReceivedContext, Task> OnTokenResponseReceived { get; set; } = context => Task.FromResult(0);
+        public Func<TokenResponseMessageReceivedContext, Task> OnTokenResponseMessageReceived { get; set; } = context => Task.FromResult(0);
+
+        /// <summary>
+        /// Invoked when an IdToken has been validated and produced an AuthenticationTicket.
+        /// </summary>
+        public Func<TokenValidatedContext, Task> OnTokenValidated { get; set; } = context => Task.FromResult(0);
```

Comment:
This event has been historically used for performing any work that requires knowledge of the incoming user. The goal is signaling the fact that the token carrying identity information has been validated and its content is now available for processing. This should be consistent with all the other middlewares which validate tokens. The fact that the middleware as a whole *might* be doing more work in the OIDC case is a no-op for the implementer - if they want to do something with the access token (if present) they can do it later on.